proxy/gateway/java/gateway-remote-module-sdk/src/main/java/com/microsoft/azure/gateway/remote/MessageDeserializer.java

```diff
@@ -0,0 +1,124 @@
+    /**
+     * Deserializes the message and constructs a {@link ControlMessage}.
+     *
+     * @param messageBuffer The message content
+     * @param version The message version that should be parsed.
+     *
+     * @return
+     * @throws MessageDeserializationException If the message is malformed.
+     */
+    public RemoteMessage deserialize(ByteBuffer messageBuffer, byte version) throws MessageDeserializationException {
+        RemoteMessageType msgType = null;
+        int totalSize = 0;
+        messageBuffer.position(0);
+
+        if (messageBuffer.limit() < BASE_MESSAGE_SIZE) {
+            throw new MessageDeserializationException(
+                    String.format("Message size %s should be >= %s", messageBuffer.limit(), BASE_MESSAGE_SIZE));
+        }
+
+        byte header1 = messageBuffer.get();
+        byte header2 = messageBuffer.get();
+        if (header1 == FIRST_MESSAGE_BYTE && header2 == SECOND_MESSAGE_BYTE) {
+            byte messageVersion = messageBuffer.get();
+            if (messageVersion > version)
+                throw new MessageDeserializationException(String.format("Message version %s can not be higher than configured version %s", messageVersion, version));
+
+            byte type = messageBuffer.get();
+            msgType = RemoteMessageType.fromValue(type);
+            if (msgType == null)
+                throw new MessageDeserializationException("Invalid message type.");
+
+            totalSize = messageBuffer.getInt();
+
+            if (totalSize != messageBuffer.limit())
+                throw new MessageDeserializationException(
+                        String.format("Message size in header %s is different that actual size %s", totalSize,
+                                messageBuffer.limit()));
+
+        } else {
+            throw new MessageDeserializationException("Invalid message header.");
+        }
+
+        switch (msgType) {
+        case CREATE:
+            return this.deserializeCreateMessage(messageBuffer, totalSize);
+        case START:
+            return this.deserializeStartMessage(messageBuffer);
+        case DESTROY:
+            return this.deserializeDestroyMessage(messageBuffer);
+        default:
+            return new ControlMessage(RemoteMessageType.ERROR);
+        }
+    }
+
+    private RemoteMessage deserializeCreateMessage(ByteBuffer buffer, int totalSize)
+            throws MessageDeserializationException {
+        if (totalSize < BASE_CREATE_SIZE)
+            throw new MessageDeserializationException(
+                    String.format("Create message size %s should be >= %s", totalSize, BASE_CREATE_SIZE));
+
+        CreateMessage message = null;
+        byte version = buffer.get();
+
+        byte uriType = buffer.get();
+        int uriSize = buffer.getInt();
+
+        String id = readNullTerminatedString(buffer, uriSize);
+        DataEndpointConfig endpointConfig = new DataEndpointConfig(id, uriType);
+
+        int argsSize = buffer.getInt();
```

Comment:
Is this affected when java is 64 bit (vs. 32 bit)? The control message is always a 32 bit integer.